File: src/main/java/com/github/xumeng/java/problems/easy/problem0242/Solution0242.java
package com.github.xumeng.java.problems.easy.problem0242;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.Set;

public class Solution0242 {

    public boolean isAnagram(String s, String t) {
        if (s.length() != t.length()) {
            return false;
        }
        Map<Character, Integer> charactersOfS = countCharacters(s);
        Map<Character, Integer> charactersOfT = countCharacters(t);
        int size1 = charactersOfS.size();
        int size2 = charactersOfT.size();
        if (size1 != size2) {
            return false;
        }
        Set<Character> characters = charactersOfS.keySet();
        for (Character character : characters) {
            Integer countS = charactersOfS.get(character);
            Integer countT = charactersOfT.get(character);
            if (!Objects.equals(countS, countT)) {
                return false;
            }
        }
        return true;
    }

    private Map<Character, Integer> countCharacters(String s) {
        Map<Character, Integer> charactersMap = new HashMap<>();
        int length = s.length();
        for (int i = 0; i < length; i++) {
            char character = s.charAt(i);
            Integer count = charactersMap.get(character);
            if (count != null) {
                count++;
            } else {
                count = 1;
            }
            charactersMap.put(character, count);
        }
        return charactersMap;
    }

}

File: src/main/kotlin/com/github/xumeng/kotlin/problems/easy/problem0242/Solution0242.kt
package com.github.xumeng.kotlin.problems.easy.problem0242

class Solution0242 {

    fun isAnagram(s: String, t: String): Boolean {
        if (s.length != t.length) {
            return false
        }
        val characterOfS: Map<Char, Int> = countCharacters(s)
        val characterOfT: Map<Char, Int> = countCharacters(t)
        val size1: Int = characterOfS.size
        val size2: Int = characterOfT.size
        if (size1 != size2) {
            return false
        }
        val characters: Set<Char> = characterOfS.keys
        characters.forEach { character ->
            val countS = characterOfS[character]
            val countT = characterOfT[character]
            if (countS != countT) {
                return false
            }
        }
        return true
    }

    private fun countCharacters(str: String): Map<Char, Int> {
        val charactersMap: MutableMap<Char, Int> = mutableMapOf()
        val length = str.length
        var i = 0
        while (i < length) {
            val character = str[i]
            var count: Int? = charactersMap[character]
            if (count != null) {
                count++
            } else {
                count = 1
            }
            charactersMap[character] = count
            i++
        }
        return charactersMap
    }

}